pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1449,21 +1449,28 @@ private ByteBuf processMessageChunk(ByteBuf compressedPayload, MessageMetadata m
         // discard message if chunk is out-of-order
         if (chunkedMsgCtx == null || chunkedMsgCtx.chunkedMsgBuffer == null
                 || msgMetadata.getChunkId() != (chunkedMsgCtx.lastChunkedMessageId + 1)) {
-            // means we lost the first chunk: should never happen
-            log.info("Received unexpected chunk messageId {}, last-chunk-id{}, chunkId = {}", msgId,
-                    (chunkedMsgCtx != null ? chunkedMsgCtx.lastChunkedMessageId : null), msgMetadata.getChunkId());
-            if (chunkedMsgCtx != null) {
-                if (chunkedMsgCtx.chunkedMsgBuffer != null) {
-                    ReferenceCountUtil.safeRelease(chunkedMsgCtx.chunkedMsgBuffer);
+            // Filter duplicated chunks instead of discard it.
+            if (chunkedMsgCtx == null || msgMetadata.getChunkId() <= chunkedMsgCtx.lastChunkedMessageId) {
+                log.warn("[{}] Receive a repeated chunk messageId {}, last-chunk-id{}, chunkId = {}",
+                        msgMetadata.getProducerName(), chunkedMsgCtx == null ? null
+                                : chunkedMsgCtx.lastChunkedMessageId, msgId, msgMetadata.getChunkId());
+            } else {
+                // means we lost the first chunk: should never happen
```

Comment:
We need to release these resources: buffer and chunkedMessagesMap for this case
`msgMetadata.getChunkId() <= chunkedMsgCtx.lastChunkedMessageId` too, when expired, `expireTimeOfIncompleteChunkedMessageMillis > 0
                    && System.currentTimeMillis() > (msgMetadata.getPublishTime().`

Reply:
No, I think we should not.

Reply:
Can you explain why we should not? If we don't, don't we waste resources from the incomplete chunked msgs?

Reply:
The reason why we have an expiration policy before is because rewind is called every time we receive an out-of-order chunk. In this case, if the chunk message itself is wrong, the message will be received in an infinite loop.

Reply:
We have ignored the chunks whose `msgMetadata.getChunkId() <= chunkedMsgCtx.lastChunkedMessageId`, so it will not caught in a loop。